kvserver/rangefeed: add rangefeed level cleanup callback

wenyihu6 opened the pull request:
This patch adds RegisterRangefeedCleanUp to BufferedStream interface. Rangefeed
level can register a cleanup callback that will be invoked after
BufferedStream.Disconnect is called. Note that it might not be invoked
immediately during SendBufferedError.

Buffered registration has a dedicated goroutine to do cleanup. Unbuffered
registration does not have a dedicated goroutine, so it relies on BufferedSender
to invoke the callback when it disconnects.

Part of: https://github.com/cockroachdb/cockroach/issues/129814
Release note: none

wenyihu6 committed:
kvserver/rangefeed: move helpers to processor_helpers_test

This patch moves the helper functions in processor_test.go to
processor_helpers_test.go.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129230)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver/rangefeed: rename Send to SendUnbuffered for Stream

This patch renames Send to SendUnbuffered in Stream to distinguish
it from SendBuffered future patches are going to introduce.

Part of: #129813
Release note: none

wenyihu6 committed:
kvserver/rangefeed: rename SendIsThreadSafe to SendUnbufferedIsThreadSafe for Stream

This patch renames SendIsThreadSafe to SendUnbufferedIsThreadSafe
in Stream.

Part of: #129813
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add rangefeed level callback

This patch adds RegisterRangefeedCleanUp to BufferedStream interface. Rangefeed
level can register a cleanup callback that will be invoked after
BufferedStream.Disconnect is called. Note that it might not be invoked
immediately during SendBufferedError.

Buffered registration has a dedicated goroutine to do cleanup. Unbuffered
registration does not have a dedicated goroutine, so it relies on BufferedSender
to invoke the callback when it disconnects.

Part of: #129814
Release note: none

wenyihu6 commented:
We previously talked about returning a cleanup callback all the way back to nodes.MuxRangefeed. However, when I was implementing the changes, I found this approach less appealing. It requires every layer in the stack to handle the additional callback and forces other tests that use any functions in this stack to handle extra callback. Putting this function in the BufferedStream interface seems like a more straightforward solution. Wdyt?

wenyihu6 commented:
Closing - as we are splitting things out based on our new integration branch: https://github.com/cockroachdb/cockroach/pull/125872.
